packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java

```diff
@@ -5,42 +5,207 @@
 package io.flutter.plugins.sharedpreferences;
 
 import android.content.Context;
+import android.content.SharedPreferences;
+import android.util.Base64;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
 import io.flutter.embedding.engine.plugins.FlutterPlugin;
 import io.flutter.plugin.common.BinaryMessenger;
-import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugins.sharedpreferences.Messages.SharedPreferencesApi;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 /** SharedPreferencesPlugin */
-public class SharedPreferencesPlugin implements FlutterPlugin {
-  private static final String CHANNEL_NAME = "plugins.flutter.io/shared_preferences_android";
-  private MethodChannel channel;
-  private MethodCallHandlerImpl handler;
+public class SharedPreferencesPlugin implements FlutterPlugin, SharedPreferencesApi {
+  static final String TAG = "SharedPreferencesPlugin";
+
+  SharedPreferencesListEncoder listEncoder;
+
+  private static final String SHARED_PREFERENCES_NAME = "FlutterSharedPreferences";
+
+  // Fun fact: The following is a base64 encoding of the string "This is the prefix for a list."
+  private static final String LIST_IDENTIFIER = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIGxpc3Qu";
+  private static final String BIG_INTEGER_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBCaWdJbnRlZ2Vy";
+  private static final String DOUBLE_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBEb3VibGUu";
+
+  private android.content.SharedPreferences preferences;
+
+  public SharedPreferencesPlugin() {
+    this(new ListEncoder());
+  }
+
+  @VisibleForTesting
+  SharedPreferencesPlugin(SharedPreferencesListEncoder testListEncoder) {
+    listEncoder = testListEncoder;
+  }
 
   @SuppressWarnings("deprecation")
-  public static void registerWith(io.flutter.plugin.common.PluginRegistry.Registrar registrar) {
+  public static void registerWith(
+      @NonNull io.flutter.plugin.common.PluginRegistry.Registrar registrar) {
     final SharedPreferencesPlugin plugin = new SharedPreferencesPlugin();
-    plugin.setupChannel(registrar.messenger(), registrar.context());
+    plugin.setUp(registrar.messenger(), registrar.context());
+  }
+
+  private void setUp(@NonNull BinaryMessenger messenger, @NonNull Context context) {
```

Comment:
You've marked `context` as `@NonNull`, but are calling it with `null` in `onDetachedFromEngine`.